test/replication-luatest/gh_5158_invalid_recovery_from_xlog_test.lua

```diff
@@ -0,0 +1,88 @@
+local t = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+local fio = require('fio')
+
+local g = t.group('gh_5158')
+
+g.before_each(function(cg)
+    cg.cluster = cluster:new({})
+    cg.master = cg.cluster:build_and_add_server{
+        alias = 'master',
+        box_cfg = {
+            checkpoint_count = 5,
+        },
+    }
+    cg.replica = cg.cluster:build_and_add_server{
+        alias = 'replica',
+        box_cfg = {
+            replication = {
+                server.build_instance_uri('master'),
+            },
+        },
+    }
+    cg.cluster:start()
+end)
+
+g.after_each(function(cg)
+    cg.cluster:drop()
+end)
+
+g.test_invalid_recovery_from_xlog = function(cg)
+    cg.replica:stop()
+    t.helpers.retrying({}, function()
+        t.assert_str_matches(cg.master:exec(function()
+            return box.info.replication[2].downstream["status"]
+        end), 'stopped')
+    end)
+
+    cg.master:exec(function()
+        box.schema.space.create("test")
+        box.snapshot()
+        box.space.test:create_index("pk")
+        box.snapshot()
+    end)
+
+    cg.replica:start()
+
+    -- Wait until everything is replicated from the master to the replica
+    t.helpers.retrying({}, function()
+        cg.replica:wait_vclock_of(cg.master)
+    end)
+
+    -- Delete all *.xlogs on the replica
+    cg.replica:stop()
+    local glob = fio.pathjoin(cg.replica.workdir, '*.xlog')
+    local xlogs = fio.glob(glob)
+    for _, file in pairs(xlogs) do fio.unlink(file) end
+
+    -- We start the replica without replication so that the replica
+    -- does not catch up to the maximum vclock. Next, set
+    -- ERRINJ_WAL_DELAY_COUNTDOWN. And then we replicate.
+    -- The replica acknowledges that it has received part of the data
+    -- due to ERRINJ_WAL_DELAY_COUNTDOWN. As a result, on the intermediate
+    -- xlog on which the assert is fired (signature < prev_signature),
+    -- the GC is started.
+    --
+    -- A batch from WAL can immediately come to the replica. Accordingly,
+    -- ERRINJ_WAL_DELAY_COUNTDOWN stops the entire batch (the test
+    -- freezes), because the entire batch has arrived and the vclock
+    -- will catch up to the maximum value. Therefore, it is necessary
+    -- to use wal_queue_max_size
```

Comment:
Nice explanation of what's happening below.
Please, also mention why `wal_queue_max_size` is exactly 96, and why `replication_sync_timeout = 0.1` is needed.

Reply:
Done